mr/src/main/java/org/apache/iceberg/mr/hive/HiveSchemaConverter.java

```diff
@@ -0,0 +1,118 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.mr.hive;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.hadoop.hive.serde2.typeinfo.DecimalTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.ListTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.MapTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.PrimitiveTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.StructTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+class HiveSchemaConverter {
+  private int id;
+
+  private HiveSchemaConverter() {
+    id = 0;
+  }
+
+  static Schema convert(List<String> names, List<TypeInfo> typeInfos) {
+    HiveSchemaConverter converter = new HiveSchemaConverter();
+    return new Schema(converter.convertInternal(names, typeInfos));
+  }
+
+  List<Types.NestedField> convertInternal(List<String> names, List<TypeInfo> typeInfos) {
+    List<Types.NestedField> result = new ArrayList<>(names.size());
+    for (int i = 0; i < names.size(); ++i) {
+      result.add(Types.NestedField.optional(id++, names.get(i), convert(typeInfos.get(i))));
+    }
+
+    return result;
+  }
+
+  Type convert(TypeInfo typeInfo) {
+    switch (typeInfo.getCategory()) {
+      case PRIMITIVE:
+        switch (((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory()) {
+          case FLOAT:
+            return Types.FloatType.get();
+          case DOUBLE:
+            return Types.DoubleType.get();
+          case BOOLEAN:
+            return Types.BooleanType.get();
+          case BYTE:
+          case SHORT:
+            throw new IllegalArgumentException("Unsupported Hive type (" +
+                ((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory() +
+                ") for Iceberg tables. Consider using INT/INTEGER type instead.");
+          case INT:
+            return Types.IntegerType.get();
+          case LONG:
+            return Types.LongType.get();
+          case BINARY:
+            return Types.BinaryType.get();
+          case CHAR:
+          case VARCHAR:
+            throw new IllegalArgumentException("Unsupported Hive type (" +
+                ((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory() +
+                ") for Iceberg tables. Consider using STRING type instead.");
```

Comment:
I would be fine mapping these to string.

Reply:
Let's discuss this on he dev list where we are talking about the schema mapping.
If we decide that the Iceberg schema is the master and we always convert from there to Hive schema, then we can relax the 1-on-1 mapping restriction, and convert multiple Iceberg types to a single Hive type.